docs/design.rst

```diff
@@ -0,0 +1,403 @@
+the resulting message in the messages queue.
+
+Running this process in a task guarantees that control frames are processed
+promptly. Without such a task, ``websockets`` would depend on the application
+to drive the connection by having exactly one coroutine awaiting
+:meth:`~protocol.WebSocketCommonProtocol.recv()` at any time. While this
+happens naturally in many use cases, it cannot be relied upon.
+
+:meth:`~protocol.WebSocketCommonProtocol.recv()` fetches the next message from
+the messages queue, with some complexity added for handling termination
+correctly.
+
+Sending data
+............
+
+The right side of the diagram shows how ``websockets`` sends data.
+
+:meth:`~protocol.WebSocketCommonProtocol.send()` writes a single data frame
+containing the message. Fragmentation isn't supported at this time.
+
+:meth:`~protocol.WebSocketCommonProtocol.ping()` writes a ping frame, then
+stores and returns a :class:`~asyncio.Future` which will be completed when a
+matching pong frame is received.
+
+:meth:`~protocol.WebSocketCommonProtocol.pong()` writes a pong frame.
+
+:meth:`~protocol.WebSocketCommonProtocol.close()` writes a close frame and
+waits for the connection to terminate.
+
+Outgoing data is written to a :class:`~asyncio.StreamWriter` in order to
+implement flow control and provide backpressure from the TCP connection.
+
+.. _closing-handshake:
+
+Closing handshake
+.................
+
+When the other side of the connection initiates the closing handshake,
+:meth:`~protocol.WebSocketCommonProtocol.read_message()` receives a close
+frame while in the ``OPEN`` state. It moves to the ``CLOSING`` state, sends a
+close frame, and returns ``None``, causing
+:attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` to terminate.
+
+When this side of the connection initiates the closing handshake with
+:meth:`~protocol.WebSocketCommonProtocol.close()`, it moves to the ``CLOSING``
+state and sends a close frame. When the other side sends a close frame,
+:meth:`~protocol.WebSocketCommonProtocol.read_message()` receives it in the
+``CLOSING`` state and returns ``None``, also causing
+:attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` to terminate.
+
+If the other side doesn't send a close frame within the connection's timeout,
+:meth:`~protocol.WebSocketCommonProtocol.close()` cancels
+:attr:`~protocol.WebSocketCommonProtocol.transfer_data_task`, which has the
+same effect.
+
+Then ``websockets`` terminates the TCP connection.
+
+
+.. _connection-termination:
+
+Connection termination
+----------------------
+
+:attr:`~protocol.WebSocketCommonProtocol.close_connection_task`, which is
+started when the WebSocket connection is established, is responsible for
+eventually closing the TCP connection.
+
+First :attr:`~protocol.WebSocketCommonProtocol.close_connection_task` waits
+for :attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` to terminate,
+which may happen as a result of:
```

Comment:
I haven't reviewed all of PR #267, but I'll stop here because this seems central to the PR.

It seems like it is unnecessarily complicating things to create `close_connection_task` in advance to wait for `transfer_data_task` to terminate. Why not just use the normal control flow of a `try` block that awaits on `transfer_data_task`? That seems like the more natural and obvious thing to do.

Also, the fact that `close_connection()` needs to have an `after_handshake` boolean flag to track where in the code the method is being called seems to support this. It seems like it is compensating for what could be handled more naturally by the normal control flow of `try` blocks.

Reply:
Yes, you're correct, this is the crux of the PR.

The reason for having these two tasks is the following. I want to be able to cancel `transfer_data_task` — which is how the closing handshake times out — without risking cancelling `close_connection_task`.

If I did:
```
try:
    # transfer data
finally:
    # close connection
```
it would be more difficult to guarantee that cancellation can happen in the transfer data block but not in the close connection block.

I start `close_connection_task` as soon as the opening handshake completes rather than when `transfer_data_task` exits so I can await `close_connection_task` without worrying whether it exists.

The `after_handshake` is related to the opening handshake; it allows reusing the same code for closing the TCP connection if something fails during the TCP handshake (that's the `after_handshake=False` path). I tried creating two functions and calling the appropriate one depending on whether the opening handshake has succeeded or not. To me no version looks significantly better than the other. If you have a preference for two separate functions, I'm happy to oblige :-)

Reply:
You already merged the change, but I still don't see why all this complexity is needed. There are *lots* of options to consider within the "one task at a time" approach. It is the idiomatic approach. Have you tried them or discussed them if you needed help?

> it would be more difficult to guarantee that cancellation can happen in the transfer data block but not in the close connection block.

As I said above, there are many options, but why not just expose an API that lets you cancel only the transfer data block and not allow more general cancellation?

Alternatively, you could shield the close connection block from cancellation and/or provide a different way to cancel the close connection block if you want to support canceling that.

> I start close_connection_task as soon as the opening handshake completes rather than when transfer_data_task exits so I can await close_connection_task without worrying whether it exists.

But the normal asyncio approach is simply to await a task when you're creating it, by proceeding sequentially through the code. This problem only exists when you set up the approach of creating tasks in advance before needing them.

Reply:
Yes, I merged because I wanted to make further improvements to the documentation without creating conflicts — I'm working on them right now.

The main reason why I ended up with the two-task design is that it feels _less complex_ to me.

By separating the TCP layer and the WebSocket layer, I was able to reason about what is expected and what isn't — especially exceptions and cancellation — and write it down (see docs/design.rst). Only a handful of things can happen to each task. As a consequence I'm much more confident in the general logic.

I have to admit that I never trusted in the robustness of the previous implementation (which was the third or fourth iteration already) when it came to closing connections. I'm unsure how concurrent calls to close() or fail_connection() worked. There's a long story of changes and refactorings in this area. I never had a clear mental model of what guaranteed connections would be closed.

That's why #112 remained open for so long and why I was able to fix it today.

Also supporting this change is the fact that it fixed #240 as a side-effect.

That said, I'm not opposed to moving back to a single task model if we manage to maintain the same guarantees of correctness. Perhaps it will be easier to write a provably correct single task implementation after understanding the constraints with the simpler two-task implementation. Seeing the corresponding changes in design.rst will help us double-check the reasoning.

Reply:
There's exactly four references to `close_connection_task` in the code base. If you'd like to merge it back into `transfer_data_task` while maintaining the same behavior, go for it! We'll merge it if it doesn't look more complex or more difficult to check for correctness.

Reply:
Also, I don't know how this is currently addressed, but I think it's important to have a clear understanding of the _API_ you want to support and expose around things like cancellation and closing connections before doing refactoring. Is this currently documented? If you're clear on the API you want and what should be supported, then discussing and reasoning about the implementation is simpler.

The API should also be documented separately from the implementation details. Otherwise, it's not clear what's what.

I would also hope that the two tasks are an implementation detail and not part of the public API. Otherwise, it won't be changeable without breaking things for the end-user.

Reply:
The public API I want to support is what shows up on this page — mainly from docstrings: http://websockets.readthedocs.io/en/stable/api.html

The policy wrt. cancellation isn't sufficiently spelt out, that's a good point. I think the natural expectation is "cancelling any coroutine raises CancelledError and websockets doesn't break apart". You pointed out earlier the close() swallowed CancelledError; I fixed that :-)

Then there are things "don't leak TCP connections that could take a couple minutes to time out, if not more." which aren't usually in API docs but which I care a lot about.

Reply:
Cf. #278.

Reply:
> I think the natural expectation is "cancelling any coroutine raises CancelledError and websockets doesn't break apart".

But there is a little more to it because you also need to specify behaviorally what occurs when the CancelledError is raised (e.g. does cleanup also occur?). I think the edge cases that might be harder / take more work to document are things like coroutines that call more than coroutine inside in sequence which has already been mentioned (e.g. both transferring data and closing the connection). What do you want to happen there? Maybe there is a general policy that can be used to describe the expectation across the library, or maybe it needs to be documented per function in the docstring (though it should be without referencing internal implementation details).